rig/utils/contexts.py

```diff
@@ -167,16 +167,25 @@ def __init__(self, context_arguments, stack=None):
         """
         self.context_arguments = dict(context_arguments)
         self.stack = stack
+        self.before_closing = list()
 
     def update(self, updates):
         """Update the arguments contained within this context."""
         self.context_arguments.update(updates)
 
+    def before_close(self, fn):
+        """Call the given function before this context is exited."""
+        self.before_closing.append(fn)
+
     def __enter__(self):
         # Add this context object to the stack
         self.stack.append(self)
 
     def __exit__(self, exception_type, exception_value, traceback):
+        # Call all the passed functions before closing the context
+        for fn in self.before_closing:
```

Comment:
This and the stack-poping could be in a try: finally: block to leave things in a consistent state if a before-closing function fails.